thewalrus/reference.py

```diff
@@ -278,3 +282,121 @@ def hafnian(M, loop=False):
         tot_sum = tot_sum + result
 
     return tot_sum
+
+
+def mapper(x, objects):
+    """Helper function to turn a permutation and bistring into an element of rpmp.
+
+    Args:
+        x (tuple): tuple containing a permutation and a bistring
+        objects (list): list objects to permute
+
+    Returns:
+        tuple: permuted objects
+    """
+    (perm, bit) = x
+    m = len(bit)
+    Blist = [list(range(m)), list(range(m, 2 * m))]
+    for i, j in enumerate(bit):
+        if int(j):
+            (Blist[0][i], Blist[1][i]) = (Blist[1][i], Blist[0][i])
+    Blist = [Blist[0][i] for i in tuple((0,) + perm)] + [Blist[1][i] for i in tuple((0,) + perm)]
+    dico_list = {j: i + 1 for i, j in enumerate(Blist)}
+    new_mapping_list = {
+        objects[dico_list[i] - 1]: objects[dico_list[j] - 1]
+        for i, j in zip(list(range(0, m - 1)) + [m], list(range(m + 1, 2 * m)) + [m - 1])
+    }
+    return tuple(new_mapping_list.items())
+
+
+def bitstrings(n):
+    """Returns the bistrings from 0 to n/2
+
+    Args:
+        n (int): Twice the highest bitstring value.
+
+    Returns:
+        (iterator): An iterable of all bistrings.
+    """
+    for binary in map("".join, product("01", repeat=n - 1)):
+        yield "0" + binary
+
+
+def rpmp(s):
+    """Generates the restricted set of perfect matchings matching permutations.
+
+    Args:
+        s (tuple): tuple of labels to be used
+
+    Returns:
+        generator: the set of restricted perfect matching permutations of the tuple ``s``
+    """
+    m = len(s) // 2
+
+    def local_mapper(x):
+        """Helper function to define a local mapper based on the symbols s
+        Args:
+            x (iterable): object to be mapped
+        """
+        return mapper(x, s)
+
+    for i in product(permutations(range(1, m)), bitstrings(m)):
+        yield local_mapper(i)
+
+
+def splitter(elem):
+    """Takes an element from the restricted perfect matching permutations (rpmp) and returns all the associated elements in the restricted single pair matchings (rspm)
+
+    Args:
+        elem (tuple): tuple representing an element of rpmp
+
+    Returns:
+        (iterator): all the associated elements in rspm
+    """
+    num_elem = len(elem)
+    net = [elem]
+    for i in range(num_elem):
+        left = (elem[j] for j in range(i))
+        middle_left = ((elem[i][0], elem[i][0]),)
+        middle_right = ((elem[i][1], elem[i][1]),)
+        right = (elem[j] for j in range(i + 1, num_elem))
+        net.append(tuple(middle_right) + tuple(right) + tuple(left) + tuple(middle_left))
+    for i in net:
+        yield i
```

Comment:
In this piece of code there are two generators and two tuples but then all of them get converted into a tuple in order to join them :( this might go counter the idea of the generators because you'll end up keeping all those numbers in memory. There's a solution though: you can use [`itertools.chain`](https://docs.python.org/3/library/itertools.html#itertools.chain) to chain generators and iterables into a big happy family of lazily evaluated iterators
```suggestion
def splitter(element):
    """Takes an element from the restricted perfect matching permutations (rpmp) and returns all the associated elements in the restricted single pair matchings (rspm)

    Args:
        elem (tuple): tuple representing an element of rpmp

    Returns:
        (iterator): all the associated elements in rspm
    """
    num_element = len(element)
    net = [element]
    for i, elem in enumerate(element):
        left = (elem[j] for j in range(i))
        middle_left = ((elem[0], elem[0]),)
        middle_right = ((elem[1], elem[1]),)
        right = (elem[j] for j in range(i + 1, num_elem))
        net = chain(net, middle_right, right, left, middle_left)

    for i in net:
        yield i
```
Note: I didn't check if this actually works 🙈

Reply:
At least I tried...